src/objective/rank_objective.hpp

```diff
@@ -253,9 +306,86 @@ class LambdarankNDCG : public RankingObjective {
     }
   }
 
+  void InitPositionBiasesAndGradients() {
+    i_biases_pow_.resize(truncation_level_);
+    j_biases_pow_.resize(truncation_level_);
+    i_costs_.resize(truncation_level_);
+    j_costs_.resize(truncation_level_);
+
+    for (int i = 0; i < truncation_level_; ++i) {
+      // init position biases
+      i_biases_pow_[i] = 1.0f;
+      j_biases_pow_[i] = 1.0f;
+
+      // init position gradients
+      i_costs_[i] = 0.0f;
+      j_costs_[i] = 0.0f;
+    }
+
+    // init gradient buffers for gathering results across threads
+    for (int i = 0; i < num_threads_; i++) {
+      i_costs_buffer_.emplace_back(truncation_level_, 0.0f);
+      j_costs_buffer_.emplace_back(truncation_level_, 0.0f);
+    }
+  }
+
+  void UpdatePositionBiasesAndGradients() const {
+    // accumulate the parallel results
+    for (int i = 0; i < num_threads_; i++) {
+      for (int j = 0; j < truncation_level_; j++) {
+        i_costs_[j] += i_costs_buffer_[i][j];
+        j_costs_[j] += j_costs_buffer_[i][j];
+      }
+    }
+
+    for (int i = 0; i < num_threads_; i++) {
+      for (int j = 0; j < truncation_level_; j++) {
+        // clear buffer for next run
+        i_costs_buffer_[i][j] = 0.0f;
+        j_costs_buffer_[i][j] = 0.0f;
+      }
+    }
+
+    for (int i = 0; i < truncation_level_; i++) {
+      // Update bias
+      i_biases_pow_[i] = pow(i_costs_[i] / i_costs_[0], position_bias_regularizer);
+      j_biases_pow_[i] = pow(j_costs_[i] / j_costs_[0], position_bias_regularizer);
+    }
+
+    LogDebugPositionBiases();
```

Comment:
while i found being able to track evolution and end point of position bias adjustment values useful in debug mode, happy to remove this since it doesn't have any new feature benefit during normal usage.

Reply:
Yes. Please remove this or wrap the debug code with `#ifdef DEBUG` macro.